File: lab1a/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)

const bufferSize = 10
const threadCount = 3

type DataEntry struct {
	Name string `json:"name"`
	Sugar float32 `json:"sugar"`
	Criteria int `json:"criteria"`
}

func processEntry(entry DataEntry, outputEntries *SortedList, outputMutex *sync.Mutex) {
	time.Sleep(time.Millisecond * 100 + time.Millisecond * time.Duration(entry.Criteria))
	if entry.Sugar > float32(entry.Criteria) {
		outputMutex.Lock()
		fmt.Println("Output:", entry)
		outputEntries.Append(entry)
		outputMutex.Unlock()
	}
}

func processEntries(monitor *LimitedBuffer, outputEntries *SortedList, outputMutex *sync.Mutex, waitGroup *sync.WaitGroup) {
	for monitor.ReserveEntry() {
		entry := monitor.Remove()
		fmt.Println("Started to filter:", entry)
		processEntry(entry, outputEntries, outputMutex)
		fmt.Println("Finished to filter:", entry)
	}
	waitGroup.Done()
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage:", os.Args[0], "<data-file> <output-file>")
		os.Exit(-1)
	}

	dataFilename := os.Args[1]
	outputFilename := os.Args[2]
	fileData, err := os.ReadFile(dataFilename)
	if err != nil {
		log.Fatal(err)
	}

	entries := []DataEntry{}
	err = json.Unmarshal(fileData, &entries)
	if err != nil {
		log.Fatal(err)
	}

	outputList := makeSortedList(len(entries))
	outputMutex := sync.Mutex{}
	waitGroup := sync.WaitGroup{}

	monitor := makeLimitedBuffer(bufferSize)
	monitor.estimatedCount = len(entries)

	for i := 0; i < threadCount; i++ {
		go processEntries(&monitor, &outputList, &outputMutex, &waitGroup)
		waitGroup.Add(1)
	}

	for _, entry := range entries {
		monitor.Insert(entry);
	}
	waitGroup.Wait()

	outputFile, err := os.OpenFile(outputFilename, os.O_TRUNC | os.O_CREATE | os.O_WRONLY, 0644)
	if err != nil {
		log.Fatal(err)
	}
	defer outputFile.Close()

	outputListBytes, err := json.MarshalIndent(outputList.GetSlice(), "", "\t")
	if err != nil {
		log.Fatal(err)
	}
	outputFile.Write(outputListBytes)

	fmt.Println("Finished")
}

File: lab1a/cpp/main.cpp
#include <algorithm>
#include <chrono>
#include <thread>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

#include "json.hpp"

using namespace std;
using json = nlohmann::json;

const int bufferSize = 10;
const int threadCount = 1;

#include "common.h"
#include "limited_buffer.cpp"

void processEntry(DataEntry *entry, vector<DataEntry> *entries, mutex *outputMutex) {
	this_thread::sleep_for(chrono::milliseconds(100 + entry->criteria));
	if (entry->sugar > entry->criteria) {
		outputMutex->lock();
		cout << "Output: " << entry->name << endl;
		// TODO: make this a sorted add
		entries->push_back(*entry);
		sort(entries->begin(), entries->end(), [](DataEntry &a, DataEntry &b) {
				return a.sugar < b.sugar;
		});
		outputMutex->unlock();
	}
}

void processEntries(LimitedBuffer *monitor, vector<DataEntry> *outputList, mutex *outputMutex) {
	while (monitor->reserve_entry()) {
		DataEntry entry = monitor->remove();
		cout << "Started to filter: " << entry.name << endl;
		processEntry(&entry, outputList, outputMutex);
		cout << "Finished to filter: " << entry.name << endl;
	}
}

int main(int argc, char **argv) {
	if (argc != 3) {
		cout << "Usage: " << argv[0] << " <data-file> <output-file>" << endl;
		return -1;
	}

	char *inputPath = argv[1];
	char *outputPath = argv[2];

	std::ifstream f(inputPath);
	json data = json::parse(f);

	vector<DataEntry> entries;
	for (auto it : data) {
		entries.push_back({
				.name = it["name"],
				.sugar = it["sugar"],
				.criteria = it["criteria"],
		});
	}

	vector<DataEntry> outputList;
	mutex outputListMutex;

	LimitedBuffer monitor(bufferSize);
	monitor.estimated_count = entries.size();

	vector<thread> threads;
	for (int i = 0; i < threadCount; i++) {
		threads.push_back(thread(processEntries, &monitor, &outputList, &outputListMutex));
	}

	for (int i = 0; i < entries.size(); i++) {
		monitor.insert(entries[i]);
	}
	for (int i = 0; i < threads.size(); i++) {
		threads[i].join();
	}

	std::ofstream outputFile(outputPath);
	outputFile << "[" << endl;
	for (int i = 0; i < outputList.size(); i++) {
		DataEntry *entry = &outputList[i];
		json json_entry;
		json_entry["name"] = entry->name;
		json_entry["sugar"] = entry->sugar;
		json_entry["criteria"] = entry->criteria;

		std::string json_str = json_entry.dump();
		outputFile << "\t" << json_str;
		if (i < outputList.size()-1) {
			outputFile << ",";
		}
		outputFile << endl;
	}
	outputFile << "]" << endl;

	cout << "Finished" << endl;
	return 0;
}
